module/zfs/arc.c

```diff
@@ -6050,6 +6052,56 @@ arc_remove_prune_callback(arc_prune_t *p)
 	kmem_free(p, sizeof (*p));
 }
 
+/*
+ * Helper function for arc_prune_async() it is responsible for safely
+ * handling the execution of a registered arc_prune_func_t.
+ */
+static void
+arc_prune_task(void *ptr)
+{
+	arc_prune_t *ap = (arc_prune_t *)ptr;
+	arc_prune_func_t *func = ap->p_pfunc;
+
+	if (func != NULL)
+		func(ap->p_adjust, ap->p_private);
+
+	zfs_refcount_remove(&ap->p_refcnt, func);
+}
+
+/*
+ * Notify registered consumers they must drop holds on a portion of the ARC
+ * buffers they reference.  This provides a mechanism to ensure the ARC can
+ * honor the metadata limit and reclaim otherwise pinned ARC buffers.
+ *
+ * This operation is performed asynchronously so it may be safely called
+ * in the context of the arc_reclaim_thread().  A reference is taken here
+ * for each registered arc_prune_t and the arc_prune_task() is responsible
+ * for releasing it once the registered arc_prune_func_t has completed.
+ */
+static void
+arc_prune_async(uint64_t adjust)
+{
+	arc_prune_t *ap;
+
+	mutex_enter(&arc_prune_mtx);
+	for (ap = list_head(&arc_prune_list); ap != NULL;
+	    ap = list_next(&arc_prune_list, ap)) {
+
+		if (zfs_refcount_count(&ap->p_refcnt) >= 2)
```

Comment:
If arc_prune_task() is about to return, but did not yet decrement the refcount, then here we will not enqueue another task even though we might want to. I am not sure if it is a serious problem in practice, and of course the Linux implementation already had this bug. OTOH, it should be straightforward to use arc_prune_mtx to synchronize arc_prune_async() and arc_prune_task().

Reply:
There is no right answer what to do with next request while previous is running, since the new prune request size depends on how much the old one still has to do. So I don't see a problem there.  arc_prune_mtx can not cover the arc_prune_task(), since those may run in parallel on Linux.

Reply:
arc_prune_mutex does not need to (and indeed shouldn't) cover the callback in arc_prune_task(), it can just be used to synchronize access to the refcount and `ap->p_adjust`. But yes I don't think there's a "right" behaviour here, so ok.